src/core/MIGRATION.md

```diff
@@ -1115,6 +1116,164 @@ class MyPlugin {
+      }
+    });
+  }
+```
+
+- register route handler for GET request to 'my-app/path/{id}' path
+```ts
+import { schema, TypeOf } from '@kbn/config-schema';
+import { Router } from 'src/core/server';
+const router = new Router('my-app');
+
+const validate = {
+  params: schema.object({
+    id: schema.string(),
+  }),
+};
+
+router.get(
+  {
+    path: 'path/{id}',
+    validate
+  },
+  async (request, response) => {
+    const data = await findObject(request.params.id);
+    if (!data) return response.notFound();
+    return response.ok(data, {
+      headers: {
+        'content-type': 'application/json'
+      }
+    });
+  }
+);
+```
+#### What data I can send back with a response?
+You have several options to create a response utilizing `response`:
+1. create a successful response. Supported types of response body are:
+- `undefined`, no content to send.
+- `string`, send text
+- `JSON`, send JSON object, HTTP server will throw if given object is not valid (has circular references, for example)
+- `Stream` send data stream
+- `Buffer` send binary stream
+```js
+return response.ok(undefined);
+return response.ok('ack');
+return response.ok({ id: '1' });
+return response.ok(Buffer.from(...););
+
+const stream = new Stream.PassThrough();
+fs.createReadStream('./file').pipe(stream);
+return res.ok(stream);
+```
+HTTP headers are configurable via response factory parameter `options`.
+
+```js
+return response.ok({ id: '1' }, {
+  headers: {
+    'content-type': 'application/json'
+  }
+});
+```
+2. create redirect response. Redirection URL is configures via 'Location' header.
+```js
+return response.redirected('The document has moved', {
+  headers: {
+    location: '/new-url',
+  },
+});
+```
+3. create error response. You may pass an error message to the client, where error message can be:
+- `string` send message text
+- `Error` send the message text of given Error object.
+- `{ message: string | Error, meta: {data: Record<string, any>, ...} }` - send message text and attach additional error metadata.
+```js
+return response.unauthorized('User has no access to the requested resource.', {
+  headers: {
+    'WWW-Authenticate': 'challenge',
+  }
+})
+return response.badRequest();
+return response.badRequest('validation error');
+
+try {
+  // ...
+} catch(error){
+ return response.badRequest(error);
+}
+
+return response.badRequest({
+  message: 'validation error',
+  meta: {
+    data: {
+      requestBody: request.body,
+      failedFields: validationResult
+    },
+  }
+});
+
+try {
+  // ...
+} catch(error){
+ return response.badRequest({
+  message: error,
+  meta: {
+    data: {
+      requestBody: request.body,
+    },
+  }
+});
+}
+
+```
+4. create a custom response. It might happen that `response factory` doesn't cover your use case and you want to specify HTTP response status code as well.
+```js
+return response.custom('ok', {
+  statusCode: 201,
+  headers: {
+    location: '/created-url'
+  }
+})
+```
```

Comment:
What about testing? Could you provide an example of how to test these handler methods?

Reply:
Testing is not relevant as I moved the docs closer to the code. 
We can provide a separate doc recommending test patterns for NP.
Although I don't it's common for plugins to test their endpoints in insolation. Usually, It's a part of integration/functional tests.

Reply:
> Although I don't it's common for plugins to test their endpoints in insolation.

Heh this really made upgrading Hapi very hard last time. IMO, endpoints should be tested with decent units so there can be fewer integrations tests needed. Anyways, I'm fine with not doing that here in this PR.